### ISSUE TITLE
[RFC] Backstage User Identity should have consistent User ID in case of multiple authentication providers

### ISSUE THREAD

**Issue by OrkoHunter** [backstage/backstage#7611]
## Need

TL;DR: The User ID obtained via Identity API is not consistent for a user when the Backstage instance has multiple authentication providers setup. This makes it difficult for plugins relying heavily on the logged in user to find a way to get consistent ID for the same user.

---

When a user logs in to Backstage using any [authentication provider](https://backstage.io/docs/auth/), the Backstage identity is created and set in the session - specifically a user ID, some profile info and a token containing entity claims, etc. This information is accessible via the [Identity API](https://backstage.io/docs/reference/core-plugin-api.identityapi). By default, the user ID is the string before `@company.com` in case of email based auth providers (e.g. google), or the username in case of github, gitlab, etc. But Backstage integrators can customize the format of user IDs by writing [Sign In Resolvers](https://backstage.io/docs/auth/identity-resolver).

It’s not rare for Backstage instances to have more than one Sign In providers. For example, in a company, users log in via Google for their “primary” identity across plugins but also log in via GitHub to supply some plugins a scoped GitHub token and authenticate against GitHub API requests from Backstage.

The `userId` stored in Backstage Identity is set based on which authentication provider the user chose to log in first. The subsequent logins using other authentication providers will have no effect on the user ID. [Source](https://github.com/backstage/backstage/blob/03d306945ee91239b6d859a24a7e7a3d8632cd08/packages/core-app-api/src/app/AppIdentity.ts#L69-L80). Hence depending on the order of sign in, users will be assigned different user IDs. This will cause problems in plugins who rely on the logged in user's identity and treat user ID to be unique and consistent.

For example: the backstage app in my company has two sign in providers - Google and GitHub. I (as a Backstage user) have the email address `himanshum@company.com` in my company and my username is `OrkoHunter` on GitHub.

Scenario 1: I sign in via Google first and then via GitHub - my `userId` is going to be `himanshum`.
Scenario 2: I sign in via GitHub first and then via Google - my `userId` is going to be `OrkoHunter`.

## Proposal

Along with one `userId`, the provider-specific user IDs and profile info should also be set in the backstage identity.

Currently there is only one `userId` and `profile` stored in the identity store.

```tsx
export class AppIdentity implements IdentityApi {
  // ...
  private userId?: string;
  private profile?: ProfileInfo;
}
```

Adding something like a new `providers` key would allow storing multiple `userId` and profile info per provider. An initial suggestion

```tsx
// Source of truth is in core-plugin-api/src/apis/definition/auth.ts
type AuthProviders = "google" | "github" | "okta" | "gitlab" | "auth0"; // etc.

type ProvidersInfo = {
  [provider in AuthProviders]: {
    userId: string;
    profile: ProfileInfo;
  };
};

export class AppIdentity implements IdentityApi {
  // ...
  private userId?: string;
  private profile?: ProfileInfo;
  private providers?: ProvidersInfo;
}
```

With this change, if we revisit the scenario 1 - I sign in via Google first and then via GitHub - the identity api would return something like this

```js
// Backstage Identity in case of multiple sign in providers
{
  userId: 'himanshum',
  profile: {
    email: 'himanshum@company.com',
    displayName: 'Himanshu Mishra',
    picture: 'https://storage.google.com/himanshum.png',
  },
  providers: {
    google: {
      userId: 'himanshum',
      profile: {
        email: 'himanshum@company.com',
        displayName: 'Himanshu Mishra',
        picture: 'https://storage.google.com/himanshum.png',
      },
    },
    github: {
      userId: 'OrkoHunter',
      profile: {
        email: 'himanshum@orkohunter.net',
        displayName: 'Himanshu Mishra',
        picture: 'https://github.com/OrkoHunter.png',
      },
    }
  }
}
```

This would allow plugins to pick a userId from a specific identity provider and expect it to always remain consistent for the same user.

Auth0 solves a similar problem with a similar [`identities` key](https://auth0.com/docs/users/user-account-linking#how-it-works).

## Alternative

An alternative solution could be in case of multiple auth providers can be to allow backstage integrators to pick a “primary” auth provider as the identity provider which is used to set user id, profile info, etc. None of the other auth providers would then be used to get the userId.

An example would be - if I (as a Backstage integrator) have configured Google to be the "primary" identity provider for the users of my Backstage instance - only their Google login will be used to set the `userId`. And if users log in via GitHub or other identity providers, the `userId` will not be set.

I think this alternative of having a "primary identity provider" can be done along with the expansion of identity to include user info from all providers.

Open for thoughts!

**Comment by OrkoHunter** [comment]
For transparency, I'm planning to start the implementation proposed here. Hopefully the Pull Request with code will be able to gauge more comments than this issue RFC. 😁

**Comment by freben** [comment]
Sorry I've been slow getting to this!

With the huge caveat that I may be misunderstanding the proposal:

I think it's important to note that the concept of "identity" (as in, the act of signing in and claiming who the current user is) in the frontend is a completely different concept to the authorization flows (as in, the act of acquiring some tokens that may or may not be attached to some real user). They are distinct subsystems and do not interact with each other. Confusingly, they are implemented in large parts using the exact same _backend_ mechanisms, but that's a choice made more from a code sharing / simplicity of configuration standpoint, not from a similarity of frontend purpose standpoint.

There is only ever one frontend sign-in. It's the act of unblocking the `SignInPage`, and making the rest of the app render. The result of signing in should be some loose concept of a "primary" identity, something that the org at hand considers to be what best tells who the real person is. This is often something like an LDAP username. A typical org would have exactly one sign-in possibility, and no need for multiple entries in the `IdentityApi` because there is no way of having more than one active "session" (identity).

Then separately from that, there is the auth API. There, you can indeed have multiple active auth "sessions" at the same time, as in, you can hold multiple access/id tokens for third party resources. But those do not affect your identity. You can find all of these identities and corresponding IDs and profiles separately.

Let me know if I responded coherently and correctly here or if I'm entirely barking up the wrong tree :)

**Comment by OrkoHunter** [comment]
@freben I think you are quite coherent here. And I got a new insight for the first time. Up until now, I was under the impression that if you add multiple authentication providers (typically the list on the Settings page -> Authentication providers), you would also add their corresponding sign-in providers to the `SignInPage` as well.

And all the issues mentioned here arrive when you have multiple providers in the `SignInPage` - because in that case the user's `userId` in the identity API depends on which sign in provider they chose first.

**Comment by OrkoHunter** [comment]
Notes from a voice chat with @freben et al 

* Our recommendation for organizations is to only have one Sign In provider on their `SignInPage` even though they have multiple authentication providers enabled for third party access i.e. GitHub, Gitlab, etc.
* Even though orgs can have multiple SignInProviders, they can setup a [custom Sign In/Identity Resolver](https://backstage.io/docs/auth/identity-resolver) to guarantee a consistent userId for individuals.

**Comment by freben** [comment]
It may be the case that we can clarify (or slim down!) the sign in offering on both the example app and the default create-app template, if that helps.

**Comment by OrkoHunter** [comment]
> It may be the case that we can clarify (or slim down!) the sign in offering on both the main repo's example app and the default create-app template, if that helps.

💯 agreed. cc @mtlewis
